Python/Product/Debugger/Debugger/PythonProcess.cs

```diff
@@ -620,476 +535,337 @@ public PythonProcess(PythonLanguageVersion languageVersion, string exe, string a
-                    }
-                ));
+        private void OnLegacyException(object sender, LDP.ExceptionEvent e) {
+            var exc = new PythonException();
+            foreach (var item in e.data) {
+                exc.SetValue(this, item.key, item.val);
             }
-            _stoppedForException = true;
-        }
 
-        private void HandleRichException(Stream stream) {
-            var exc = new PythonException();
-            long tid = stream.ReadInt64();
-            int count = stream.ReadInt32();
-            while (--count >= 0) {
-                string key = stream.ReadString();
-                string value = stream.ReadString();
-                exc.SetValue(this, key, value);
-            }
-            if (tid != 0) {
-                Debug.WriteLine("Exception: " + exc.FormattedDescription ?? exc.ExceptionMessage ?? exc.TypeName);
-                ExceptionRaised?.Invoke(this, new ExceptionRaisedEventArgs(_threads[tid], exc));
+            if (e.threadId != 0) {
+                Debug.WriteLine("Exception: " + (exc.FormattedDescription ?? exc.ExceptionMessage ?? exc.TypeName));
+                ExceptionRaised?.Invoke(this, new ExceptionRaisedEventArgs(_threads[e.threadId], exc));
                 _stoppedForException = true;
             }
         }
 
-        private static string CommandtoString(byte[] cmd_buffer) {
-            return new string(new char[] { (char)cmd_buffer[0], (char)cmd_buffer[1], (char)cmd_buffer[2], (char)cmd_buffer[3] });
+        private void OnLegacyDetach(object sender, LDP.DetachEvent e) {
+            _process_Exited(this, EventArgs.Empty);
         }
 
-        private void HandleLast(Stream stream) {
-            DebugWriteCommand("LAST ack");
-            lock (_streamLock) {
-                stream.Write(LastAckCommandBytes);
+        private void OnLegacyLast(object sender, LDP.LastEvent e) {
+            try {
+                SendDebugRequestAsync(new LDP.LastAckRequest()).WaitAndUnwrapExceptions();
+            } catch (IOException) {
+            } catch (ObjectDisposedException) {
+                // The process waits for this request with short timeout before terminating
+                // If the process has terminated, we expect an exception
             }
         }
 
-        internal void SendStepOut(long threadId) {
-            DebugWriteCommand("StepOut");
-            lock (_streamLock) {
-                _stream.Write(StepOutCommandBytes);
-                _stream.WriteInt64(threadId);
-            }
+        private void OnLegacyInternalError(object sender, LDP.InternalErrorEvent e) {
+            // This event is sent when an internal error occurs
+            // on python side while processing a request.
+            // Request handlers should probably catch errors and send them back in the response
+            // instead of letting it propagate up.
+            Debug.Fail(e.output);
         }
```

Comment:
Looking at it more closely, my comment above is wrong.
on_request calls the specific request handler ie. on_legacyStepInto, and catches any exception and does a send_response(success=False, ...) with exception info.
The only way that an internal error would be sent is if that send_response reporting the error fails. If that fails, it's probably because there's a problem with the socket, and sending the internal error event would fail too.
So I'm thinking we just remove this whole internal error event and just do a traceback.print_exc() on the python side.

---
In reply to: [108259436](https://github.com/Microsoft/PTVS/pull/2323#discussion_r108259436) [](ancestors = 108259436,108023087,107985719)